### PR TITLE
implement index class set operations methods

### PR DESCRIPTION
**Pull request by wkdavis**
Signed-off-by: wkdavis <william.davis@worthingtonindustries.com>

### CODE REVIEW

**Comment by wkdavis** [comment]
@earowang If you have a moment to take a look at this, I'm not sure why namespace WARNINGs are being generated. I am not able to reproduce them locally.

```r
> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS  10.16

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tsibble_0.9.4  testthat_3.0.0

loaded via a namespace (and not attached):
 [1] tidyselect_1.1.0  xfun_0.18         remotes_2.2.0     purrr_0.3.4       rematch2_2.1.2    diffobj_0.3.2     rcmdcheck_1.3.3   vctrs_0.3.4       generics_0.1.0    usethis_1.6.3    
[11] rlang_0.4.8       pkgbuild_1.1.0    pillar_1.4.6      glue_1.4.2        withr_2.3.0       waldo_0.2.2       xopen_1.0.0       sessioninfo_1.1.1 lifecycle_0.2.0   stringr_1.4.0    
[21] commonmark_1.7    anytime_0.3.9     devtools_2.3.2    memoise_1.1.0     knitr_1.30        callr_3.5.1       ps_1.4.0          fansi_0.4.1       Rcpp_1.0.5        backports_1.1.10 
[31] desc_1.2.0        pkgload_1.1.0     fs_1.5.0          digest_0.6.27     stringi_1.5.3     processx_3.4.4    dplyr_1.0.2       rprojroot_1.3-2   cli_2.1.0         tools_3.6.1      
[41] magrittr_1.5      tibble_3.0.4      crayon_1.3.4      pkgconfig_2.0.3   ellipsis_0.3.1    xml2_1.3.2        prettyunits_1.1.1 lubridate_1.7.9   assertthat_0.2.1  roxygen2_7.1.1   
[51] rstudioapi_0.11   R6_2.4.1          compiler_3.6.1   
```

File: R/yearmonth.R [hunk 1]
```diff
@@ -278,6 +278,30 @@ seq.yearmonth <- function(from, to, by, length.out = NULL, along.with = NULL,
   }
 }
 
+#' @export
+union.yearmonth <- function(x, y, ...) {
+  if (!inherits(y, "yearmonth")) {
```
Comment (earowang) [inline R/yearmonth.R]:
```suggestion
  if (!is_yearmonth(y)) {
```

File: R/yearmonth.R [hunk 2]
```diff
@@ -278,6 +278,30 @@ seq.yearmonth <- function(from, to, by, length.out = NULL, along.with = NULL,
...
+    stop("'y' must be of class 'yearmonth'")
```
Comment (earowang) [inline R/yearmonth.R]:
```suggestion
    abort("'y' must be of class 'yearmonth'")
```

File: R/yearmonth.R [hunk 3]
```diff
@@ -278,6 +278,30 @@ seq.yearmonth <- function(from, to, by, length.out = NULL, along.with = NULL,
...
+  }
+  unique(vec_c(x, y))
```
Comment (earowang) [inline R/yearmonth.R]:
```suggestion
  vec_unique(vec_c(x, y))
```

File: R/yearmonth.R [hunk 4]
```diff
@@ -278,6 +278,30 @@ seq.yearmonth <- function(from, to, by, length.out = NULL, along.with = NULL,
...
+}
+
+#' @export
+intersect.yearmonth <- function(x, y, ...) {
+  if (!inherits(y, "yearmonth")) {
+    stop("'y' must be of class 'yearmonth'")
+  }
+  x[vec_in(x,y)]
```
Comment (earowang) [inline R/yearmonth.R]:
```suggestion
  vec_slice(x, [vec_in(x, y)])
```

File: R/yearmonth.R [hunk 5]
```diff
@@ -278,6 +278,30 @@ seq.yearmonth <- function(from, to, by, length.out = NULL, along.with = NULL,
...
+}
+
+#' @export
+setdiff.yearmonth <- function(x, y, ...) {
+  if (!inherits(y, "yearmonth")) {
+    stop("'y' must be of class 'yearmonth'")
+  }
+  c(x[!vec_in(x, y)], y[!vec_in(y, x)])
```
Comment (earowang) [inline R/yearmonth.R]:
```suggestion
  vec_c(x[!vec_in(x, y)], y[!vec_in(y, x)])
```

File: R/zzz.R [hunk 6]
```diff
@@ -20,5 +20,15 @@
...
+  s3_register("generics::intersect", "yearquarter")
+  s3_register("generics::intersect", "yearmonth")
+  s3_register("generics::intersect", "yearweek")
+  s3_register("generics::union", "yearquarter")
+  s3_register("generics::union", "yearmonth")
+  s3_register("generics::union", "yearweek")
+  s3_register("generics::setdiff", "yearquarter")
+  s3_register("generics::setdiff", "yearmonth")
+  s3_register("generics::setdiff", "yearweek")
+
```
Comment (earowang) [inline R/zzz.R]: You don't need to register these methods on load, since they've been exported.

**Review by earowang — COMMENTED** [review]
Thanks. Looks great! 

Can you incorporate some my reviews, to use {vctrs} more? Can you add a couple of examples in each help page and add yourself as a contributor?

**Comment by wkdavis** [comment]
@earowang 
> Can you incorporate some my reviews, to use {vctrs} more? Can you add a couple of examples in each help page and add yourself as a contributor?

I expanded the use of {vctrs} and added myself as a contributor in `DESCRIPTION`. As far as examples, there is no separate documentation for these methods because I felt that the generics (from {generics}) are sufficiently documented and include examples. However, if you'd like I can add documentation specific to the `year*` methods that includes examples. Thanks!

**Comment by earowang** [comment]
Thank you.

**Pull request merged**